activerecord/lib/active_record/connection_adapters/abstract/query_comment.rb

```diff
@@ -0,0 +1,120 @@
+          else
+            "#{sql} #{comment}"
+          end
+        end
+        sql
+      end
+
+      # Maintains a user-defined context for all queries and constructs an SQL comment by
+      # calling methods listed in :components
+      #
+      # Additional information can be pushed to the context to be referenced by methods
+      # defined in framework or application initializers.
+      module Context
+        mattr_accessor :components, :lines_to_ignore, instance_accessor: false
+        mattr_accessor :cache_query_comment, instance_accessor: false, default: false
+        thread_mattr_accessor :context, instance_accessor: false, default: {}
+        thread_mattr_accessor :cached_comment, instance_accessor: false
+        DEFAULT_LINES_TO_IGNORE_REGEX = %r{\.rvm|/ruby/gems/|vendor/|active_record|active_support|rbenv|monitor\.rb.*mon_synchronize}
+
+        class << self
+          Context.components = [:application]
+
+          def comment
+            return cached_comment if cache_query_comment && !cached_comment.nil?
+            self.cached_comment = "/*#{escape_sql_comment(comment_content)}*/"
+          end
+
+          def escape_sql_comment(str)
+            while str.include?('/*') || str.include?('*/')
+              str = str.gsub('/*', '').gsub('*/', '')
+            end
+            str
+          end
+
+          def comment_content
+            components.filter_map do |c|
+              if component_value = self.try(c)
+                "#{c}:#{component_value}"
+              end
+            end.join(",")
+          end
+
+          def set(ctx)
+            return unless ctx.is_a? Hash
+            self.context.merge!(ctx)
+            self.cached_comment = nil
+          end
+
+          def context_available?(context_name)
+            context.key?(context_name) && context[context_name].present?
+          end
+
+          def clear_comment_cache!
+            self.cached_comment = nil
+          end
+
+          # Set during Rails boot in lib/active_record/railtie.rb
+          def application
+            context[:application_name] || "active_record"
+          end
+
+          def connection_config
+            ActiveRecord::Base.connection_db_config
+          end
+
+          def socket
+            connection_config.socket
+          end
+
+          def db_host
+            connection_config.host
+          end
+
+          def database
+            connection_config.database
+          end
+
+          def line
+            self.lines_to_ignore ||= DEFAULT_LINES_TO_IGNORE_REGEX
+            last_line = caller.detect do |line|
+              line !~ lines_to_ignore
+            end
+            if last_line
+              root = if defined?(Rails.root)
+                Rails.root.to_s
+              elsif defined?(RAILS_ROOT)
+                RAILS_ROOT
+              else
+                ""
+              end
+              last_line.delete_prefix(root)
+            end
+          end
+        end
```

Comment:
Rails has this concept built-in as `BacktraceCleaner` 😁

```
backtrace_cleaner.clean(caller.lazy).first
```

(see comment above for `backtrace_cleaner` definition)